### PR TITLE
Library

### PR DESCRIPTION
**Pull request by FredTheDino**
Change the engine into a library, this is a fundamental change to the engine since the engine and the game cannot communicate at compile time. This is looking to be a gigantic patch, so early input is probably a good idea.

### CODE REVIEW

**Comment by FredTheDino** [comment]
It's not working yet, fully, but it's a good base case.

**Comment by sornas** [comment]
Looks cool, I'll try writing a small game and get a feel for how the bindings work. Just a heads up, I'm unable to compile `example.cpp` on mac (but it works fine on Linux). It stops after 20 errors and starts with a few

```
out/fog.h:305:1: error: unknown type name 'constexpr'
constexpr f32 AUDIO_DEFAULT_GAIN = 0.2;
^
out/fog.h:305:14: error: expected ';' after top level declarator
constexpr f32 AUDIO_DEFAULT_GAIN = 0.2;
             ^
             ;
out/fog.h:307:1: error: unknown type name 'constexpr'
constexpr f32 AUDIO_DEFAULT_VARIANCE = 0.01;
^
out/fog.h:307:14: error: expected ';' after top level declarator
constexpr f32 AUDIO_DEFAULT_VARIANCE = 0.01;
             ^
```
and is followed later on by some
```
example.cpp:37:22: error: expected expression
    p_s.alive_time = {1, 1};
                     ^
example.cpp:38:24: error: expected expression
    p_s.velocity_dir = {0, 2*3.14};
                       ^
```
with quite a few
```
out/fog.h:520:19: warning: 'fog_renderer_camera_smooth' has C-linkage specified, but returns user-defined type 'Camera' which is incompatible with C [-Wreturn-type-c-linkage]
FOG_IMPORT Camera fog_renderer_camera_smooth(Camera camera_a, Camera camera_b, f32 slerp);
                  ^
```
in between. The library itself compiles with `make` but I'm getting ~50 copies of the following warning (at different locations).
```
src/engine/math/block_vector.h:36:39: warning: suggest braces around initialization of subobject [-Wmissing-braces]
    Vec2i operator-() const { return {-x, -y}; }
                                      ^~~~~~
                                      {     }
```
I'm also getting this at the end:
```
src/engine/../fog_bindings.cpp:123:19: warning: 'fog_renderer_camera_lerp' has C-linkage specified, but returns user-defined type 'Renderer::Camera' which is incompatible with C [-Wreturn-type-c-linkage]
FOG_IMPORT Camera fog_renderer_camera_lerp(Camera camera_a, Camera camera_b, f32 lerp) { return Renderer::camera_lerp(camera_a, camera_b, lerp); }
                  ^
src/engine/../fog_bindings.cpp:124:19: warning: 'fog_renderer_camera_smooth' has C-linkage specified, but returns user-defined type 'Renderer::Camera' which is incompatible with C [-Wreturn-type-c-linkage]
FOG_IMPORT Camera fog_renderer_camera_smooth(Camera camera_a, Camera camera_b, f32 slerp) { return Renderer::camera_smooth(camera_a, camera_b, slerp); }
                  ^
src/engine/../fog_bindings.cpp:125:19: warning: 'fog_renderer_camera_fit' has C-linkage specified, but returns user-defined type 'Renderer::Camera' which is incompatible with C [-Wreturn-type-c-linkage]
FOG_IMPORT Camera fog_renderer_camera_fit(u32 num_points, Vec2 *points, f32 border) { return Renderer::camera_fit(num_points, points, border); }
                  ^
src/engine/../fog_bindings.cpp:143:27: warning: 'fog_renderer_create_particle_system' has C-linkage specified, but returns user-defined type 'Renderer::ParticleSystem' which is incompatible with C [-Wreturn-type-c-linkage]
FOG_IMPORT ParticleSystem fog_renderer_create_particle_system(u32 layer, u32 num_particles, Vec2 position) { return
Renderer::create_particle_system(layer, num_particles, position); }
                          ^
```
As you know, this probably (hopefully) won't affect me going forward, but I figured you'd want to know. As I said, it works fine on my Linux-machine, so it sounds to me like apple is doing something funky with the compiler. For reference:
```
$ g++ --version
Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/c++/4.2.1
Apple LLVM version 10.0.1 (clang-1001.0.46.4)
Target: x86_64-apple-darwin18.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

File: compile.sh [hunk 1]
```diff
@@ -0,0 +1,2 @@
+#!/usr/bin/bash
```
Comment (sornas) [inline compile.sh]: On the topic of mac:

```suggestion
#!/usr/bin/env bash
```

File: src/engine/asset/asset.h [hunk 2]
```diff
@@ -140,43 +142,58 @@ u64 asset_hash(const char *str) {
     return hash % 0x0FFFFFFF;
 }
 
+constexpr u64 asset_hash_comp(const char *str) {
+    u64 hash = 5351;
```
Comment (sornas) [inline src/engine/asset/asset.h:147]: Where is `5351` coming from?

File: src/engine/header.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
+// This allows the exporting of structs in a nice way, this macro does nothing
+// other than marking C++ code for export to the exporter.
+#define FOG_HIDE
+#define FOG_EXPORT
+#define FOG_EXPORT_STRUCT extern "C"
```
Comment (sornas) [inline src/engine/header.h]: IIRC `extern "C"` didn't help so I'd rather remove it and let the macro not do anything at all.

File: src/engine/header.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+// Is set when the engine is compiled as an engine.
+#define FOG_ENGINE
+
+// Bellow here are the settings for the engine
```
Comment (sornas) [inline src/engine/header.h]:
```suggestion
// Below here are the settings for the engine
```

File: src/engine/platform/input.h [hunk 5]
```diff
@@ -21,14 +22,14 @@ enum class Player {
...
+} Player;
 
-bool is_valid_player(Player p) {
+b8 is_valid_player(Player p) {
     return (p == Player::P1 || p == Player::P2 || p == Player::P3 ||
             p == Player::P4);
 }
 
-bool is(Player p, Player filter) { return (bool)((u8)p & (u8)filter); }
+b8 is(Player p, Player filter) { return (b8)((u8)p & (u8)filter); }
```
Comment (sornas) [inline src/engine/platform/input.h:32]: This could be cast directly to `b8`.

File: src/engine/platform/input.h [hunk 6]
```diff
@@ -184,29 +193,40 @@ struct Mapping {
...
     InputCode code;
-    bool pressed;
+    b8 pressed;
     f32 value;
 };
 
+///*
+// Gives back a new unique name that can be used to
+// react to inputs. If multiple names are requested,
+// the first of the names are returned, and the subsekent
```
Comment (sornas) [inline src/engine/platform/input.h]:
```suggestion
// the first of the names are returned, and the subsequent
```

File: src/engine/platform/mixer.h [hunk 7]
```diff
@@ -47,7 +48,7 @@ struct Channel {
         f32 len_seconds_target;
         f32 len_seconds_delta;
         f32 _prev_len_seconds;
-        operator bool() const {
+        operator b8() const {
```
Comment (sornas) [inline src/engine/platform/mixer.h:51]: These (two more below) should probably be exposed in the bindings in some way.

File: src/engine/renderer/particle_system.h [hunk 8]
```diff
@@ -40,35 +40,35 @@ struct Particle {
...
-    bool one_color;
-    bool one_alpha;
-    bool one_size;
-    bool drop_oldest;
+    u8 relative;
+    u8 keep_alive;
+    u8 one_color;
+    u8 one_alpha;
+    u8 one_size;
+    u8 drop_oldest;
```
Comment (sornas) [inline src/engine/renderer/particle_system.h]: Why not `b8`?

File: src/engine/util/io.cpp [hunk 9]
```diff
@@ -47,9 +47,20 @@ u32 str_len(const char *str) {
 // format string (see printf for more info). Memory
 // is allocated using the frame allocator so it is
 // valid for two frames and does not require freeing.
-char *format(const char *fmt, ...);
+char *format(const char *fmt, va_list args);
 
-char *format(const char *fmt, ...) {
+char *format(const char *fmt, va_list args) {
+    // TODO(ed): Test this function.
```
Comment (sornas) [inline src/engine/util/io.cpp:53]: boi have I got just the branch for you

File: src/engine/asset/asset.h (hunk 2 above)
Comment (FredTheDino) [reply src/engine/asset/asset.h:147]: Magical constant for a good hash number.

File: src/engine/platform/input.h (hunk 5 above)
Comment (sornas) [reply src/engine/platform/input.h:32]: OOF

Comment (FredTheDino) [reply src/engine/platform/input.h:32]: No.

**Review by sornas — APPROVED** [review]

**Pull request merged**